s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/input/AvroTransformer.java

```diff
@@ -58,6 +64,20 @@ public byte[] getValueBytes(final Object record, final String topic, final S3Sou
                 s3SourceConfig);
     }
 
+    private Stream<Object> readAvroRecordsAsStream(final IOSupplier<InputStream> inputStreamIOSupplier,
+            final DatumReader<GenericRecord> datumReader) {
+        try (DataFileStream<GenericRecord> dataFileStream = new DataFileStream<>(inputStreamIOSupplier.get(),
+                datumReader)) {
+            // Wrap DataFileStream in a Stream using a Spliterator for lazy processing
```

Comment:
I'm  trying to figure out why the dataFileStream would not be `close()`ed before any records are read!  Obviously it's not the case, but ... why?

Reply:
Now with this dataFileStream in try autocloseable block, it should be closed when the full stream with multiple avro recs are read. We had a challenge here as its a stream, placing in finally block is closing the stream after first read.
So now its looks better.

Reply:
Hope I answered :) ?

Reply:
To summarize our offline talk (it was a complicated one).

1. It looks like this might be a real issue, but it's really tricky to test with our current IntegrationTest setup.  
2. It only affects Avro, not the Parquet or Json implementation :partying_face: 
3. In certain situations, the `DataFileStream` wrapping an `S3ObjectInputStream` can _successfully_ find records when the methods are being called in this order: `open(...)`, `close()`, `hasNext() / next()`, `close()`.  

In other situations, we can see `java.io.IOException: Attempted read on closed stream` due to the misplaced `close()` call, but the job continues as work was progressing (although it never completes).

Since we're on a longer-lived feature branch, I'm going to merge this and create an KCON Jira to investigate and resolve this before it gets merged into main.